src/group.cr

```diff
@@ -0,0 +1,184 @@
+  end
+
+  # Checks to see if a group with the given group ID exists.
+  #
+  # Returns: `true` if the group exists.
+  #
+  # ```
+  # Group.exists?("root")              # => true
+  # Group.exists?("nonexistant_group") # => false
+  # ```
+  def self.exists?(gid : Int) : Bool
+    return !LibC.getgrgid(int_to_gid(gid)).null?
+  end
+
+  # Checks to see if group with a given groupname exists.
+  #
+  # Returns: `true` if the group exists.
+  #
+  # ```
+  # Group.exists?(0)     # => true
+  # Group.exists?(32766) # => false
+  # ```
+  def self.exists?(group : String | Symbol) : Bool
+    return !LibC.getgrnam(group.to_s.check_no_null_byte).null?
+  end
+
+  # Returns the group specified by the group ID.
+  # ```
+  # Group[0]
+  # ```
+  def self.[](gid : Int) : Group
+    return new(gid)
+  end
+
+  # Returns the group specified by the groupname.
+  # ```
+  # Group["root"]
+  # ```
+  def self.[](groupname : String | Symbol) : Group
+    return new(groupname)
+  end
+
+  # Initializes a group by the given group ID.
+  def initialize(gid : Int)
+    group_struct = LibC.getgrgid(Group.int_to_gid(gid))
+    raise NotFound.new("Group with gid '#{gid}', was not found.") if group_struct.null?
+
+    group_struct = group_struct.value
+    @name = String.new(group_struct.gr_name)
+    @gid = group_struct.gr_gid
+    @member_names = group_struct.gr_mem.to_slice_null_terminated(LibC::NGROUPS_MAX).to_a() { |chars| String.new(chars) }
+  end
+
+  # Initializes a group by the given groupname.
+  def initialize(groupname : String | Symbol)
+    group_struct = LibC.getgrnam(groupname.to_s.check_no_null_byte)
+    raise NotFound.new("Group with name '#{groupname.to_s}', was not found.") if group_struct.null?
+
+    group_struct = group_struct.value
+    @name = String.new(group_struct.gr_name)
+    @gid = group_struct.gr_gid
+    @member_names = group_struct.gr_mem.to_slice_null_terminated(LibC::NGROUPS_MAX).to_a() { |chars| String.new(chars) }
+  end
+
+  # Returns the groups name.
+  #
+  # ```
+  # Group[0].name # => "root"
+  # ```
+  getter(name : String)
+
+  # Returns the group ID.
+  #
+  # ```
+  # Group["root"].gid # => 0
+  # ```
+  getter(gid : UInt32)
+
+  # Returns an array of the groups members names.
+  #
+  # ```
+  # Group["root"].member_names
+  # ```
+  getter(member_names : Array(String))
+
+  # Yields the groups members names.
+  #
+  # ```
+  # Group["root"].member_names() { |name| puts names }
+  # ```
+  def member_names(&block) : Nil
+    @members.each() { |member| yield(member) }
+  end
+
+  # Returns an array of the groups members.
+  #
+  # ```
+  # Group["root"].members
+  # ```
+  def members : Array(User)
+    return @members.map() { |member| next User[member] }
+  end
+
+  # Yields the groups members.
+  #
+  # ```
+  # Group["root"].members() { |user| puts user.name }
+  # ```
+  def members(&block) : Nil
+    @members.each() { |member|
+      next yield(User[member])
```

Comment:
dito re: `next`.

multi-line blocks are typically done using `do`/`end` instead of `{`.

```cr
def each_member
  each_member_name { |name| User.new(name) }
end
```

is how i'd write this.

Actually, where did `@members` come from here? It doesn't appear to me to be defined anywhere...

Reply:
Fixed. @members was from an earlier iteration and this method isn't covered in the tests as there isn't good/consistent testing material. I added a spec that kind of works its not very good on linux though because the `root` group doesn't have any users.